wpe/src/main/glue/common/jnihelper.cpp

```diff
@@ -1,18 +1,39 @@
-#include <jnihelper.h>
-
-bool getJniEnv(JNIEnv **env) {
-    bool didAttachThread = false;
-    *env = nullptr;
-    auto get_env_result = vm->GetEnv((void**)env, JNI_VERSION_1_6);
-    if (get_env_result == JNI_EDETACHED) {
-        if (vm->AttachCurrentThread(env, NULL) == JNI_OK) {
-            didAttachThread = true;
+#include "jnihelper.h"
+
+#include <sys/prctl.h>
+
+namespace wpe {
+namespace android {
+namespace {
+
+JavaVM *g_jvm = NULL;
+
+}
+
+void InitVM(JavaVM *vm) {
+    g_jvm = vm;
+}
+
+JNIEnv *AttachCurrentThread() {
+    JNIEnv *env = nullptr;
+    jint ret = g_jvm->GetEnv(reinterpret_cast<void **>(&env), JNI_VERSION_1_6);
+    if (ret == JNI_EDETACHED || !env) {
+        JavaVMAttachArgs args;
+        args.version = JNI_VERSION_1_6;
+        args.group = nullptr;
+
+        char thread_name[16]; // 16 is the max size for android thread names
+        int err = prctl(PR_GET_NAME, thread_name);
+        if (err < 0) {
+            args.name = nullptr;
         } else {
-            throw;
+            args.name = thread_name;
         }
-    } else if (get_env_result == JNI_EVERSION) {
-        // Unsupported JNI version.
-        throw;
+
+        ret = g_jvm->AttachCurrentThread(&env, &args);
```

Comment:
Documentation is not very clear about the fact that on error the value of &env is not changed.
Maybe you should check if ret is JNI_OK or if &env is not touched on error, maybe it is not useful to store the returned value in ret.

Reply:
Fixed